### PR TITLE
fix dubbo proxyless xds connect, fix Connection closed after GOAWAY, …

### PR DESCRIPTION
**Pull request by wucheng1997**
…add error deal

## What is the purpose of the change

fix https://github.com/apache/dubbo/issues/10494
dubbo proxyless方案在一段时间后无法感知服务变化和服务节点变化，主要由于observer异常导致无法监听相关变动

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java [hunk 1]
```diff
@@ -231,6 +231,8 @@ public void onNext(DiscoveryResponse value) {
         @Override
         public void onError(Throwable t) {
             logger.error("xDS Client received error message! detail:", t);
+            StreamObserver<DiscoveryRequest> requestObserver = xdsChannel.createDeltaDiscoveryRequest(new ResponseObserver(requestId));
+            requestObserverMap.put(requestId, requestObserver);
```
Comment (AlbumenJ) [inline dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java:235]: 这里需要检查下 Throwable 的类型，同时做下重试的速率限制

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java (hunk 1 above)
Comment (wucheng1997) [reply dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java]: > 这里需要检查下 Throwable 的类型，同时做下重试的速率限制

做了一些修改，please help review

File: dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java [hunk 2]
```diff
@@ -231,6 +234,16 @@ public void onNext(DiscoveryResponse value) {
         @Override
         public void onError(Throwable t) {
             logger.error("xDS Client received error message! detail:", t);
+
+            if (t instanceof StatusRuntimeException) {
+                try {
+                    StreamObserver<DiscoveryRequest> requestObserver = xdsChannel.createDeltaDiscoveryRequest(new ResponseObserver(requestId));
+                    requestObserverMap.put(requestId, requestObserver);
```
Comment (AlbumenJ) [inline dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java:241]: 这块重建后不需要重新发起请求吗？
重试的逻辑可以整合一起管理，包括网络抖动那些

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java (hunk 2 above)
Comment (wucheng1997) [reply dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java:241]: > 这块重建后不需要重新发起请求吗？ 重试的逻辑可以整合一起管理，包括网络抖动那些

这里没有重建xdsChannel，实际上这个xdsChannel一直是有效的，出问题是因为observer在执行了onError方法之后，无法在执行onNext，导致了出现后面observer无法工作，无法感知路由变化。所以我这里新建了一个observer。等下一次与istio通信就可以正常工作。

channel的一些重试等操作在grpc中应该已经有实现。

Comment (AlbumenJ) [reply dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java:241]: 这个问题在 envoy 和 istio-agent 是怎么解决的？感觉通过报错然后重建这个方式不太好（不过onError还是要处理的）

Comment (wucheng1997) [reply dubbo-xds/src/main/java/org/apache/dubbo/registry/xds/util/protocol/AbstractProtocol.java]: https://github.com/grpc/grpc-java/issues/9522
目前看来这个observer的onError方法是在grpc层实现的，这里是没办法修改，在目前的实现方案，能做的只能是重试，envoy的实现也是new connection，我修改了一下，看一下这种实现是否合理一些。

**Review by AlbumenJ — APPROVED** [review]

**Pull request merged**